compiler/rustc_metadata/src/rmeta/decoder.rs

```diff
@@ -714,23 +728,29 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
     }
 
     fn def_kind(&self, item_id: DefIndex) -> DefKind {
-        self.root.tables.def_kind.get(self, item_id).map(|k| k.decode(self)).unwrap_or_else(|| {
-            bug!(
-                "CrateMetadata::def_kind({:?}): id not found, in crate {:?} with number {}",
-                item_id,
-                self.root.name,
-                self.cnum,
+        *self.def_kinds.borrow_mut().entry(item_id).or_insert_with(|| {
+            self.root.tables.def_kind.get(self, item_id).map(|k| k.decode(self)).unwrap_or_else(
+                || {
+                    bug!(
+                        "CrateMetadata::def_kind({:?}): id not found, in crate {:?} with number {}",
+                        item_id,
+                        self.root.name,
+                        self.cnum,
+                    )
+                },
```

Comment:
`DefKind` is very cheap to decode I think. https://github.com/rust-lang/rust/blob/c7dbe7a830100c70d59994fd940bf75bb6e39b39/compiler/rustc_hir/src/def.rs#L55-L119